File: wui/src/host-browser/host-browser.c
/* 
 *  Copyright (C) 2008  Red Hat, Inc.
 *  Written by Chris Lalancette <clalance@redhat.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdio.h>
#include <assert.h>
#include <stdlib.h>
#include <time.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/wait.h>
#include <netdb.h>
#include <arpa/inet.h>
#include <string.h>
#include <errno.h>
#include <signal.h>
#include <unistd.h>

#include <avahi-client/client.h>
#include <avahi-client/lookup.h>

#include <avahi-common/simple-watch.h>
#include <avahi-common/malloc.h>
#include <avahi-common/error.h>

#ifndef DBWRITER_PATH
#define DBWRITER_PATH "./dbwriter.rb"
#endif

static AvahiSimplePoll *simple_poll = NULL;

static void usage(void)
{
  fprintf(stderr, "Usage: host-browser [OPTIONS]\n");
  fprintf(stderr, "OPTIONS:\n\n");
  fprintf(stderr, " -d\t\tRun in daemon mode (the default)\n");
  fprintf(stderr, " -h\t\tPrint this help message\n");
  fprintf(stderr, " -n\t\tRun in interactive (non-daemon) mode (useful for debugging)\n");
  exit(1);
}

static void sig_chld(int signo)
{
  int status;

  if (waitpid(-1, &status, WNOHANG) < 0) {
    fprintf(stderr, "Error doing waitpid for child\n");
    return;
  }
}

// the function to make a daemon out of this program
static int daemonize(void)
{
  pid_t pid;

  if((pid=fork()) < 0){
    return -1;
  }
  else if (pid != 0){
    exit(0);
  }

  setsid();

  //  umask(0);

  return 0;
}

static void resolve_callback(AvahiServiceResolver *r, AvahiIfIndex interface,
			     AVAHI_GCC_UNUSED AvahiProtocol protocol,
			     AvahiResolverEvent event, const char *name,
			     const char *type, const char *domain,
			     const char *host_name, const AvahiAddress *address,
			     uint16_t port, AvahiStringList *txt,
			     AvahiLookupResultFlags flags,
			     AVAHI_GCC_UNUSED void* userdata)
{
    assert(r);

    /* Called whenever a service has been resolved successfully or timed out */

    switch (event) {
        case AVAHI_RESOLVER_FAILURE:
            break;

        case AVAHI_RESOLVER_FOUND: {
	  char a[AVAHI_ADDRESS_STR_MAX];
	  in_addr_t remote;
	  struct hostent *host;
	  char *argv[3];
	  pid_t pid;
	  int ret;
	  char *libvirt_hostname;

	  avahi_address_snprint(a, sizeof(a), address);

	  remote = inet_addr(a);
	  host = gethostbyaddr(&remote, sizeof(remote), AF_INET);
	  if (host == NULL) {
	    // we failed to resolve the address to a hostname; we'll just try
	    // with the IP address
	    libvirt_hostname = a;
	  }
	  else {
	    libvirt_hostname = host->h_name;
	  }

	  argv[0] = DBWRITER_PATH;
	  argv[1] = libvirt_hostname;
	  argv[2] = NULL;

	  pid = fork();

	  if (pid < 0) {
	    fprintf(stderr, "Failed to fork: %s\n",strerror(errno));
	  }
	  else if (pid == 0) {
	    // child
	    ret = execv(DBWRITER_PATH, argv);
	    if (ret < 0) {
	      fprintf(stderr, "Failed to exec %s: %s\n",DBWRITER_PATH,strerror(errno));
	    }
	  }
	  else {
	    // parent, do nothing; we'll catch the child exits with SIGCHLD
	  }

	  break;
        }
    }

    avahi_service_resolver_free(r);
}

static void browse_callback(AvahiServiceBrowser *b, AvahiIfIndex interface,
			    AvahiProtocol protocol, AvahiBrowserEvent event,
			    const char *name, const char *type,
			    const char *domain,
			    AVAHI_GCC_UNUSED AvahiLookupResultFlags flags,
			    void* userdata)
{    
    AvahiClient *c = userdata;
    assert(b);

    /* Called whenever a new services becomes available on the LAN or is removed from the LAN */

    switch (event) {
        case AVAHI_BROWSER_FAILURE:
            
            avahi_simple_poll_quit(simple_poll);
            return;

        case AVAHI_BROWSER_NEW:
            /* We ignore the returned resolver object. In the callback
               function we free it. If the server is terminated before
               the callback function is called the server will free
               the resolver for us. */

            if (!(avahi_service_resolver_new(c, interface, protocol, name, type, domain, AVAHI_PROTO_UNSPEC, 0, resolve_callback, c)))
                fprintf(stderr, "Failed to resolve service '%s': %s\n", name, avahi_strerror(avahi_client_errno(c)));
            
            break;

        case AVAHI_BROWSER_REMOVE:
            break;

        case AVAHI_BROWSER_ALL_FOR_NOW:
        case AVAHI_BROWSER_CACHE_EXHAUSTED:
            break;
    }
}

static void client_callback(AvahiClient *c, AvahiClientState state,
			    AVAHI_GCC_UNUSED void * userdata)
{
    assert(c);

    /* Called whenever the client or server state changes */

    if (state == AVAHI_CLIENT_FAILURE) {
        fprintf(stderr, "Server connection failure: %s\n", avahi_strerror(avahi_client_errno(c)));
        avahi_simple_poll_quit(simple_poll);
    }
}

int main(AVAHI_GCC_UNUSED int argc, AVAHI_GCC_UNUSED char *argv[])
{
    AvahiClient *client = NULL;
    AvahiServiceBrowser *sb = NULL;
    int error;
    int ret = 1;
    int daemon_mode = 1;
    int c;
    struct sigaction act;

    while ((c = getopt(argc, argv,":dhn")) != -1) {
      switch(c) {
      case 'd':
	daemon_mode = 1;
	break;
      case 'h':
	usage();
	break;
      case 'n':
	daemon_mode = 0;
	break;
      default:
	usage();
	break;
      }
    }

    if ((argc - optind) != 0) {
      usage();
    }

    if (daemon_mode) {
        daemonize();
    }

    act.sa_handler = sig_chld;
    sigemptyset(&act.sa_mask);
    act.sa_flags = SA_NOCLDSTOP;
    sigaction(SIGCHLD, &act, NULL);

    /* Allocate main loop object */
    if (!(simple_poll = avahi_simple_poll_new())) {
        fprintf(stderr, "Failed to create simple poll object.\n");
        goto fail;
    }

    /* Allocate a new client */
    client = avahi_client_new(avahi_simple_poll_get(simple_poll), 0, client_callback, NULL, &error);

    /* Check wether creating the client object succeeded */
    if (!client) {
        fprintf(stderr, "Failed to create client: %s\n", avahi_strerror(error));
        goto fail;
    }
    
    /* Create the service browser */
    if (!(sb = avahi_service_browser_new(client, AVAHI_IF_UNSPEC, AVAHI_PROTO_UNSPEC, "_libvirt._tcp", NULL, 0, browse_callback, client))) {
        fprintf(stderr, "Failed to create service browser: %s\n", avahi_strerror(avahi_client_errno(client)));
        goto fail;
    }

    /* Run the main loop */
    avahi_simple_poll_loop(simple_poll);
    
    ret = 0;
    
fail:
    
    /* Cleanup things */
    if (sb)
        avahi_service_browser_free(sb);
    
    if (client)
        avahi_client_free(client);

    if (simple_poll)
        avahi_simple_poll_free(simple_poll);

    return ret;
}

File: wui/src/host-browser/host-browser.rb
#!/usr/bin/ruby -Wall
#
# Copyright (C) 2008 Red Hat, Inc.
# Written by Darryl L. Pierce <dpierce@redhat.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; version 2 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA  02110-1301, USA.  A copy of the GNU General Public License is
# also available at http://www.gnu.org/copyleft/gpl.html.

$: << File.join(File.dirname(__FILE__), "../dutils")
$: << File.join(File.dirname(__FILE__), "../")

require 'rubygems'
require 'libvirt'
require 'dutils'

require 'socket'
require 'krb5_auth'
include Krb5Auth
require 'daemons'
include Daemonize

include Socket::Constants

$logfile = '/var/log/ovirt-wui/host-browser.log'

# +HostBrowser+ communicates with the a managed node. It retrieves specific information
# about the node and then updates the list of active nodes for the WUI.
#
class HostBrowser
    attr_accessor :logfile
    attr_accessor :keytab_dir
    attr_accessor :keytab_filename

    def initialize(session)
        @session = session
        @log_prefix = "[#{session.peeraddr[3]}] "
        @keytab_dir = '/usr/share/ipa/html/'
    end

    # Ensures the conversation starts properly.
    #
    def begin_conversation
        puts "#{@log_prefix} Begin conversation" unless defined?(TESTING)
        @session.write("HELLO?\n")

        response = @session.readline.chomp
        raise Exception.new("received #{response}, expected HELLO!") unless response == "HELLO!"
    end

    # Retrieves the mode request from the remote system.
    #
    def get_mode
        puts "#{@log_prefix} Determining the runtime mode." unless defined?(TESTING)
        @session.write("MODE?\n")
        response = @session.readline.chomp
        puts "#{@log_prefix} MODE=#{response}" unless defined?(TESTING)

        response
    end

    # Requests node information from the remote system.
    #
    def get_remote_info
        puts "#{@log_prefix} Begin remote info collection" unless defined?(TESTING)
        result = Hash.new
        result['HOSTNAME'] = @session.peeraddr[2]
        result['IPADDR']   = @session.peeraddr[3]

        @session.write("INFO?\n")

        loop do
            info = @session.readline.chomp

            puts "Received info='#{info}'"

            break if info == "ENDINFO"

            case info
            when "CPU"
                cpu = get_cpu_info
                cpu_info = result['CPUINFO']

                if(cpu_info == nil)
                    cpu_info = Array.new
                    result['CPUINFO'] = cpu_info
                end

                cpu_info << cpu
            when "NIC"
                nic = get_nic_info
                nic_info = result['NICINFO']

                if(nic_info == nil)
                    nic_info = Array.new
                    result['NICINFO'] = nic_info
                end

                nic_info << nic
            else

                raise Exception.new("ERRINFO! Excepted key=value : #{info}\n") unless info =~ /[\w]+[\s]*=[\w]/

                key, value = info.split("=")

                puts "#{@log_prefix} ::Received - #{key}:#{value}" unless defined?(TESTING)
                result[key] = value

                @session.write("ACK #{key}\n")
            end
        end

        return result
    end

    # Extracts CPU details from the managed node.
    #
    def get_cpu_info
        puts "Begin receiving CPU details"

        result = Hash.new

        @session.write("CPUINFO?\n")

        loop do
            info = @session.readline.chomp

            break if info == "ENDCPU"

            raise Exception.new("ERRINFO! Excepted key=value : #{info}\n") unless info =~ /[\w]+[\s]*=[\w]/

            key, value = info.split("=")

            puts "#{@log_prefix} ::Received - #{key}:#{value}" unless defined?(TESTING)
            result[key] = value

            @session.write("ACK #{key}\n")
        end

        @session.write("ACK CPU\n");

        return result
    end

    # Extracts NIC details from the managed node.
    #
    def get_nic_info
        puts "Begin receiving NIC details"

        result = Hash.new

        @session.write("NICINFO?\n")

        loop do
            info = @session.readline.chomp

            break if info == "ENDNIC"

            raise Exception.new("ERRINFO! Excepted key=value : #{info}\n") unless info =~ /[\w]+[\s]*=[\w]/

            key, value = info.split("=")

            puts "#{@log_prefix} ::Received - #{key}:#{value}" unless defined?(TESTING)
            result[key] = value

            @session.write("ACK #{key}\n")
        end

        @session.write("ACK NIC\n");

        return result
    end

    # Writes the supplied host information to the database.
    #
    def write_host_info(host_info)
        ensure_present(host_info,'HOSTNAME')
        ensure_present(host_info,'ARCH')
        ensure_present(host_info,'MEMSIZE')
        ensure_present(host_info,'CPUINFO')
        ensure_present(host_info,'NICINFO')

        cpu_info = host_info['CPUINFO']
        nic_info = host_info['NICINFO']

        cpu_info.each do |cpu|
            ensure_present(cpu,'CPUNUM')
            ensure_present(cpu,'CORENUM')
            ensure_present(cpu,'NUMCORES')
            ensure_present(cpu,'VENDOR')
            ensure_present(cpu,'MODEL')
            ensure_present(cpu,'FAMILY')
            ensure_present(cpu,'CPUIDLVL')
            ensure_present(cpu,'SPEED')
            ensure_present(cpu,'CACHE')
            ensure_present(cpu,'FLAGS')
        end

        puts "Searching for existing host record..." unless defined?(TESTING)
        host = Host.find(:first, :conditions => ["hostname = ?", host_info['HOSTNAME']])

        if host == nil
            begin
                puts "Creating a new record for #{host_info['HOSTNAME']}..." unless defined?(TESTING)

                host = Host.create(
                    "uuid"            => host_info['UUID'],
                    "hostname"        => host_info['HOSTNAME'],
                    "hypervisor_type" => host_info['HYPERVISOR_TYPE'],
                    "arch"            => host_info['ARCH'],
                    "memory"          => host_info['MEMSIZE'],
                    "is_disabled"     => 0,
                    "hardware_pool"   => HardwarePool.get_default_pool,
                    # Let host-status mark it available when it
                    # successfully connects to it via libvirt.
                    "state"           => Host::STATE_UNAVAILABLE)
            rescue Exception => error
                puts "Error while creating record: #{error.message}" unless defined?(TESTING)
            end
        else
            host.uuid         = host_info['UUID']
            host.hostname     = host_info['HOSTNAME']
            host.arch         = host_info['ARCH']
            host.memory       = host_info['MEMSIZE']
        end

        # delete an existing CPUs and create new ones based on the data
        puts "Deleting any existing CPUs"
        Cpu.delete_all(['host_id = ?', host.id])

        puts "Saving new CPU records"
        cpu_info.collect do |cpu|
            detail = Cpu.new(
                "cpu_number"      => cpu['CPUNUM'],
                "core_number"     => cpu['CORENUM]'],
                "number_of_cores" => cpu['NUMCORES'],
                "vendor"          => cpu['VENDOR'],
                "model"           => cpu['MODEL'],
                "family"          => cpu['FAMILY'],
                "cpuid_level"     => cpu['CPUIDLVL'],
                "speed"           => cpu['SPEED'],
                "cache"           => cpu['CACHE'],
                "flags"           => cpu['FLAGS'])

            host.cpus << detail
         end

        # Update the NIC details for this host:
        # -if the NIC exists, then update the IP address
        # -if the NIC does not exist, create it
        # -any nic not in this list is deleted

        puts "Updating NIC records for the node"
        nics = Array.new

        host.nics.collect do |nic|
            found = false

            nic_info.collect do |detail|
                # if we have a match, then update the database and remove
                # the received data to avoid creating a dupe later
                if detail['MAC'] == nic.mac
                    nic_info.delete(detail)
                end
            end

            # if the record wasn't found, then remove it from the database
            unless found
                host.nics.delete(nic)
                nic.destroy
            end
        end

        # iterate over any nics left and create new records for them.

        nic_info.collect do |nic|
            puts "Creating a new nic..."
            detail = Nic.new(
                'mac'        => nic['MAC'],
                'bandwidth'  => nic['BANDWIDTH'],
                'usage_type' => 1)

            host.nics << detail
        end

        host.save!

        return host
    end

    # Creates a keytab if one is needed, returning the filename.
    #
    def create_keytab(hostname, ipaddress, krb5_arg = nil)
        krb5 = krb5_arg || Krb5.new

        default_realm = krb5.get_default_realm
        libvirt_princ = 'libvirt/' + hostname + '@' + default_realm
        outfile = ipaddress + '-libvirt.tab'
        @keytab_filename = @keytab_dir + outfile

        # TODO need a way to test this portion
        unless (defined? TESTING) || File.exists?(@keytab_filename)
            # TODO replace with Kr5Auth when it supports admin actions
            puts "Writing keytab file: #{@keytab_filename}" unless defined?(TESTING)
            kadmin_local('addprinc -randkey ' + libvirt_princ)
            kadmin_local('ktadd -k ' + @keytab_filename + ' ' + libvirt_princ)

            File.chmod(0644,@keytab_filename)
        end

        hostname = `hostname -f`.chomp

        @session.write("KTAB http://#{hostname}/ipa/config/#{outfile}\n")

        response = @session.readline.chomp

        raise Exception.new("ERRINFO! No keytab acknowledgement") unless response == "ACK"
    end

    # Ends the conversation, notifying the user of the key version number.
    #
    def end_conversation
        puts "#{@log_prefix} Ending conversation" unless defined?(TESTING)

        @session.write("BYE\n");
    end

    private

    # Private method to ensure that a required field is present.
    #
    def ensure_present(info,key)
        raise Exception.new("ERROR! Missing '#{key}'...") if info[key] == nil
    end

    # Executes an external program to support the keytab function.
    #
    def kadmin_local(command)
        system("/usr/kerberos/sbin/kadmin.local -q '" + command + "'")
    end
end

def entry_point(server)
    while(session = server.accept)
        child = fork do
            remote = session.peeraddr[2]

            puts "Connected to #{remote}" unless defined?(TESTING)

            # This is needed because we just forked a new process
            # which now needs its own connection to the database.
            database_connect

            begin
                browser = HostBrowser.new(session)

                browser.begin_conversation
                case browser.get_mode
                    when "AWAKEN": browser.create_keytab(remote,session.peeraddr[3])
                    when "IDENTIFY": browser.write_host_info(browser.get_remote_info)
                end

                browser.end_conversation
            rescue Exception => error
                session.write("ERROR #{error.message}\n")
                puts "ERROR #{error.message}" unless defined?(TESTING)
            end

            puts "Disconnected from #{remote}" unless defined?(TESTING)
        end

        Process.detach(child)
    end
end

unless defined?(TESTING)
    # The main entry point.
    #
    unless ARGV[0] == "-n"
        daemonize
        # redirect output to the log
        STDOUT.reopen $logfile, 'a'
        STDERR.reopen STDOUT
    end

    server = TCPServer.new("",12120)
    entry_point(server)
end
